kafka-connector/src/main/java/com/google/pubsub/kafka/sink/CloudPubSubSinkTask.java

```diff
@@ -356,20 +357,29 @@ public void flush(Map<TopicPartition, OffsetAndMetadata> partitionOffsets) {
       if (outstandingFuturesForTopic == null) {
         continue;
       }
+      int partition = partitionOffset.getKey().partition();
       OutstandingFuturesForPartition outstandingFutures =
-          outstandingFuturesForTopic.get(partitionOffset.getKey().partition());
+          outstandingFuturesForTopic.get(partition);
       if (outstandingFutures == null) {
         continue;
       }
       try {
-        ApiFutures.allAsList(outstandingFutures.futures).get();
+        // Only wait for partition to complete if the clush hasn't already failed.
+        if (maybeException == null) {
```

Comment:
I'm a little confused by this. If we have multiple partitions, one of which fails, we dont get the future for the rest but do clear the tracking? The publish could still be outstanding for the other partitions at the point and so we just wont care about it?

Is the expectation that once `flush` fails for any partition, all partitions will be `put` from the last committed offset?

Reply:
Yea, either the preCommit/flush succeeds for all the passed in partitions or fails for all of them and the task seeks back to the last committed offsets. https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java#L410

At that point, we know anything we were tracking for those partitions will get replayed.

Reply:
Thanks for confirming.